src/main/java/org/veupathdb/service/eda/ds/core/AbstractPlugin.java

```diff
@@ -611,4 +659,155 @@ public String getRBinListAsString(List<String> labels) {
+    
+    return studyVocabsListAsRString;
+  }
+
+  public String getRStudyVocabsAsString(List<DynamicDataSpecImpl> dataSpecs) {
+    String studyVocabListRString = "veupathUtils::StudySpecificVocabulariesByVariableList(S4Vectors::SimpleList(";
+    boolean first = true;
+
+    for (int i = 0; i < dataSpecs.size(); i++) {
+      String studyVocabRString = getRStudyVocabsAsString(dataSpecs.get(i));
+      if (first) {
+        studyVocabListRString = studyVocabListRString + studyVocabRString;
+        first = false;
+      } else {
+        studyVocabListRString = studyVocabListRString + "," + studyVocabRString;
+      }
+    }
+    
+    return studyVocabListRString + "))";
+  }
+
+  public boolean hasStudyDependentVocabulary(DynamicDataSpecImpl dataSpec) {
+    PluginUtil util = getUtil();
+
+    if (dataSpec.isCollectionSpec()) {
+      return util.getHasStudyDependentVocabulary(dataSpec.getCollectionSpec());
+    } else if (dataSpec.isVariableSpec()) {
+      return util.getHasStudyDependentVocabulary(dataSpec.getVariableSpec());
+    } else {
+      return false;
+    }
+  }
+
+  public List<DynamicDataSpecImpl> findDataSpecsWithStudyDependentVocabs(Map<String, DynamicDataSpecImpl> dataSpecs) {
+    List<DynamicDataSpecImpl> dataSpecsWithStudyDependentVocabs = new ArrayList<>();
+
+    for (DynamicDataSpecImpl dataSpec : dataSpecs.values()) {
+      if (hasStudyDependentVocabulary(dataSpec)) {
+        dataSpecsWithStudyDependentVocabs.add(dataSpec);
+      } 
+    }
+            
+    return dataSpecsWithStudyDependentVocabs;
+  }
+
+  public List<DynamicDataSpecImpl> findVariableSpecsWithStudyDependentVocabs(Map<String, VariableSpec> varSpecs) {
+    Map<String, DynamicDataSpecImpl> dataSpecs = varMapToDynamicDataMap(varSpecs);
+
+    return findDataSpecsWithStudyDependentVocabs(dataSpecs);
+  }
+
+  public List<DynamicDataSpecImpl> findCollectionSpecsWithStudyDependentVocabs(Map<String, CollectionSpec> collectionSpecs) {
+    Map<String, DynamicDataSpecImpl> dataSpecs = collectionMapToDynamicDataMap(collectionSpecs);
+
+    return findDataSpecsWithStudyDependentVocabs(dataSpecs);
+  }
+
+  public String getDynamicDataSpecEntityId(DynamicDataSpecImpl dataSpec) {
+    if (dataSpec.isCollectionSpec()) {
+      return dataSpec.getCollectionSpec().getEntityId();
+    } else if (dataSpec.isVariableSpec()) {
+      return dataSpec.getVariableSpec().getEntityId();
+    } else {
+      return null;
+    }
+  }
+
+  public String getDynamicDataSpecId(DynamicDataSpecImpl dataSpec) {
+    if (dataSpec.isCollectionSpec()) {
+      return dataSpec.getCollectionSpec().getCollectionId();
+    } else if (dataSpec.isVariableSpec()) {
+      return dataSpec.getVariableSpec().getVariableId();
+    } else {
+      return null;
+    }
+  }
+
+  public String getRMegastudyAsString(String compressedDataHandle, Map<String, DynamicDataSpecImpl> dataSpecs) {
```

Comment:
I'm a little slow at following this. Does this function require certain named inputs to be defined in the R connection that's going to evaluate it?

Can we pass in the R connection and validate that it has all of the prerequisite variables defined?

Reply:
It needs the tabular data from the merge service and the special vocabs from subsetting loaded, as well as a map of how the variables in the tabular data play in the plot. It puts together a string that represents a r cmd to send to rserve, and that string contains references to those dependencies. If those things aren't present in the r process trying to run the cmd, then rserve will put up an error.